### PR TITLE
Add and delete sprint participant

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:21433)](https://staging.teamyapp.com/teams/107/tasks/21433)

### CODE REVIEW

File: core/service/sprint.go [hunk 1]
```diff
@@ -611,6 +611,79 @@ func (s Sprint) DeleteSprint(ct context.Context, sprintID uint64) (entity.Sprint
...
+				return internalErr
+			}
+
+			sprint, internalErr := s.sprintDao.FindSprintByIDWithTx(ct, tx, sprintID)
+			if internalErr != nil {
+				return internalErr
+			}
+
+			sprintLength := sprint.EndAt.UTC().Sub(sprint.StartAt.UTC())
+			numOfWeeks := sprintLength / timePerWeek
```
Comment (naturetime10) [inline core/service/sprint.go:629]: There are some tasks assigned to this person in this sprint already?

File: core/service/sprint.go [hunk 2]
```diff
@@ -611,6 +611,79 @@ func (s Sprint) DeleteSprint(ct context.Context, sprintID uint64) (entity.Sprint
...
+			teamMember, internalErr = s.teamMemberDao.FindTeamMemberWithTx(ct, tx, teamID, userID)
+			if internalErr != nil {
+				return internalErr
+			}
+
+			sprint, internalErr := s.sprintDao.FindSprintByIDWithTx(ct, tx, sprintID)
+			if internalErr != nil {
+				return internalErr
+			}
+
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
```

File: core/service/sprint.go [hunk 3]
```diff
@@ -611,6 +611,79 @@ func (s Sprint) DeleteSprint(ct context.Context, sprintID uint64) (entity.Sprint
...
+			deleteSprintParticipantMutation := mutation.NewDeleteSprintParticipant(
+				s.logger,
+				s.stateSyncer,
+				s.sprintParticipantDao,
+				s.sprintDao,
+				participant.UserID,
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
				userID,
```

File: core/service/sprint.go [hunk 4]
```diff
@@ -611,6 +611,79 @@ func (s Sprint) DeleteSprint(ct context.Context, sprintID uint64) (entity.Sprint
...
+			sprint, internalErr := s.sprintDao.FindSprintByIDWithTx(ct, tx, sprintID)
+			if internalErr != nil {
+				return internalErr
+			}
+
+			participant, internalErr := s.sprintParticipantDao.FindParticipantWithTx(ct, tx, sprint.ID, userID)
+			if internalErr != nil {
+				return internalErr
+			}
+
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/sprint.go (hunk 1 above)
Comment (duanxianbo) [reply core/service/sprint.go:629]: what does that change?

Comment (duanxianbo) [reply core/service/sprint.go:629]: oh, you mean to subtract the bandwidth with those task efforts?

File: core/service/sprint.go [hunk 5]
```diff
@@ -611,6 +611,69 @@ func (s Sprint) DeleteSprint(ct context.Context, sprintID uint64) (entity.Sprint
...
+			sprintLength := sprint.EndAt.UTC().Sub(sprint.StartAt.UTC())
+			numOfWeeks := sprintLength / timePerWeek
+
+			totalBandwidth := teamMember.WeeklyBandwidth * numOfWeeks
+			participant := entity.SprintParticipant{
+				SprintID:        sprint.ID,
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
				SprintID:        sprintID,
```

File: core/service/sprint.go [hunk 6]
```diff
@@ -611,6 +611,69 @@ func (s Sprint) DeleteSprint(ct context.Context, sprintID uint64) (entity.Sprint
...
+				UserID:          teamMember.UserID,
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
				UserID:          userID,
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/sprint.go [hunk 7]
```diff
@@ -629,10 +628,29 @@ func (s Sprint) AddTeamMemberToSprint(ct context.Context, sprintID uint64, teamI
...
+			if internalErr != nil {
+				return internalErr
+			}
+
+			tasks, internalErr := s.taskDao.FindTasksByIDsWithTx(ct, tx, taskIDs)
+			if internalErr != nil {
+				return internalErr
+			}
+
+			var subtractedBandwidth time.Duration = 0
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
			var usedBandwidth time.Duration = 0
```

File: core/service/sprint.go [hunk 8]
```diff
@@ -629,10 +628,29 @@ func (s Sprint) AddTeamMemberToSprint(ct context.Context, sprintID uint64, teamI
...
+			for _, task := range tasks {
+				if task.OwnerUserID != &userID {
+					continue
+				}
+
+				subtractedBandwidth += *task.Effort
+			}
+
+			totalBandwidth := teamMember.WeeklyBandwidth*numOfWeeks - subtractedBandwidth
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
			totalBandwidth := teamMember.WeeklyBandwidth*numOfWeeks
```

File: core/service/sprint.go [hunk 9]
```diff
@@ -629,10 +628,29 @@ func (s Sprint) AddTeamMemberToSprint(ct context.Context, sprintID uint64, teamI
...
 			participant := entity.SprintParticipant{
-				SprintID:        sprint.ID,
-				UserID:          teamMember.UserID,
+				SprintID:        sprintID,
+				UserID:          userID,
 				TotalBandwidth:  totalBandwidth,
 				UnusedBandwidth: totalBandwidth,
```
Comment (naturetime10) [inline core/service/sprint.go]: Can we create a variable for unusedBandwidth?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**